WalletWasabi/Extensions/MemoryExtensions.cs

```diff
@@ -0,0 +1,50 @@
+using Nito.AsyncEx;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+	public static class MemoryExtensions
+	{
+		private static Dictionary<object, AsyncLock> AsyncLocks { get; } = new Dictionary<object, AsyncLock>();
+
+		private static object AsyncLocksLock { get; } = new object();
+
+		public static async Task<TItem> AtomicGetOrCreateAsync<TItem>(this IMemoryCache cache, object key, Func<ICacheEntry, Task<TItem>> factory)
+		{
+			if (cache.TryGetValue(key, out TItem value))
+			{
+				return value;
+			}
+
+			AsyncLock asyncLock;
+			lock (AsyncLocksLock)
+			{
+				// Cleanup the evicted asynclocks first.
+				foreach (var toRemove in AsyncLocks.Keys.Where(x => !cache.TryGetValue(x, out _)).ToList())
```

Comment:
This is not correct. The `AsyncLocks` dictionary is `static` and here we are removing netries based on the keys of the `cache` instance received as argument. So, this will work well if we have only one cache instances.

Reply:
The following test fails because the cache is bypassed and the `ExpensiveComputation` is called 3 times instead of 2:

```c#
		[Fact]
		public async Task MultiplesCaches()
		{
			var invoked = 0;
			string ExpensiveComputation(string argument)
			{
				invoked++;
				return "Hello " + argument;
			}

			var cache1 = new MemoryCache(new MemoryCacheOptions());
			var result0 = await cache1.AtomicGetOrCreateAsync(
				"the-same-key",
				(entry) =>
				{
					entry.SetAbsoluteExpiration(TimeSpan.FromMilliseconds(10));
					return Task.FromResult(ExpensiveComputation("World!"));
				}
			);

			var cache2 = new MemoryCache(new MemoryCacheOptions());
			var result1 = await cache2.AtomicGetOrCreateAsync(
				"the-same-other-key",
				(entry) =>
				{
					entry.SetAbsoluteExpiration(TimeSpan.FromMilliseconds(10));
					return Task.FromResult(ExpensiveComputation("Lurking Wife!"));
				}
			);

			var result2 = await cache1.AtomicGetOrCreateAsync(
				"the-same-key",
				(entry) =>
				{
					entry.SetAbsoluteExpiration(TimeSpan.FromMilliseconds(10));
					return Task.FromResult(ExpensiveComputation("World!"));
				}
			);
			Assert.Equal(result0, result2);
			Assert.Equal(1, invoked);
		}
```

Reply:
Thanks, fixed. Test passes now.